engines/python/src/main/java/ai/djl/python/engine/Connection.java

```diff
@@ -127,46 +127,26 @@ static String[] getPythonStartCmd(PyEnv pyEnv, Model model, int workerId, int po
         if (pyEnv.isMpiMode()) {
             String cudaDevices = getVisibleDevices(workerId, tensorParallelDegree);
             logger.info("Set CUDA_VISIBLE_DEVICES={}", cudaDevices);
-            String[] args = new String[36];
-            args[0] = "mpirun";
-            args[1] = "-N";
+            pyEnv.addEnv("CUDA_VISIBLE_DEVICES", cudaDevices);
+            // TODO: explicit set PYTHON_EXEC for torchrun
+            // pyEnv.addEnv("PYTHON_EXEC", "$PYTHONPATH");
+            String[] args = new String[15];
+            args[0] = "torchrun";
```

Comment:
Can we just try `mpirun` but add `--bind-to none` option?

I'm running that locally now to see how it compares.

Reply:
For `torchrun`, the `pyEnv` is the same object in all the workers, and they are all started in parallel. So there's a race condition on modifing the `pyEnv` here. I'm seeing them all finish updating `pyEnv` before any of them create `envp` from it and launch the process. So they all end up launching with the last worker's value of `CUDA_VISIBLE_DEVICES`, the way this is written.

The easiest fixes I can think of are to copy `pyEnv`, or to have `getEnvironmentVariables` take an optional list of additional variable overrides and put it there. Can't have multiple workers updating `pyEnv` in parallel.

Reply:
Ok so, I ran some tests:
* between `torchrun` and `mpirun --bind-to none`, performance was about the same in 4 workers, TP 1 case, for EleutherAI/gpt-j-6b on g5.12xlarge, generating 128 new tokens, 40 iterations, concurrency 4,  DeepSpeed backend w/smoothquant. `mpirun` without `--bind-to none` has degraded performance for this test.
* EleutherAI/gpt-neox-20b on g5.12xlarge, TP4 (1 worker), I ran with `mpirun`, `torchrun`, and `mpirun --bind-to none`.
  - `mpirun` and `mpirun --bind-to none` performed the same as far as I could tell
  - `torchrun` was slightly slower, but consistently

So it looks like `mpirun --bind-to none` is the way to go.